2019-03-Autumn/0_arrays_sorting/README.md

```diff
@@ -70,21 +85,103 @@ TODO :bug:
 
 TODO :bug:
 
-### 2. PROBLEM 2 TODO :bug:
+### 2. Minimum Swaps to Group All 1's Together
 
-Source: TODO :bug:
+Source: [GeeksforGeeks](https://www.geeksforgeeks.org/minimum-swaps-required-group-1s-together/)
 
 #### Naive/Simple Solution
 
-TODO :bug:
+##### Steps
+
+Our goal is to (somehow) swap the 1's in the array so that all of the 1's are next to each other. So, we can think in reverse and figure out the thing we are looking for is the subarray with length *`1's count`* that already contains the most 1's (aka the destination). The count of 0's in that subarray will then be the minimum swaps required.
+<!-- First, we count the total number of 1's in the array. Suppose the count is *x*, we then count the number of 1's in each subarray of length *x*. The minimum of that will be our answer. -->
+
+##### Implementation
+
+```python
+def min_swaps(l):
+    # count total number of 1's
+    total_count = 0
+    for num in l:
+        if num == 1:
+            total_count += 1
+    
+    max_one_count = 0
+    # iterate thru `len(l) - total_count` subarrays
+    for i in range(len(l) - total_count + 1):
+        
+        one_count = 0
+
+        # count number of 1's in a subarray
+        for j in range(total_count):
+            if l[i + j] == 1:
+                one_count += 1
+        
+        if one_count > max_one_count:
+            max_one_count = one_count
+
+    # count of 0's is the actual answer
+    return total_count - max_one_count
+```
 
 #### Optimal Solution
 
-TODO :bug:
+##### Steps
+
+Optimize the naive solution using the **sliding window** approach. We first count the number of 1's in the first subarray. Then we just advance the right and left side of the window, adding the right element to `running_count` and subtracting the left element from `running_count` followed by a check for max. This approach improves by cutting down the number of times of inspecting each element of the array.
```

Comment:
Add a note about the complexity here (why the optimal solution takes time complexity from `O(N^2)` to `O(N)`).